### PR TITLE
Generalise Poseidon gadget over `WIDTH`, `RATE`

### LINKED ISSUE #221: Generalise Poseidon gadget for different `WIDTH, RATE`s

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes #221.

Based on #224.

### CODE REVIEW

**Comment by str4d** [comment]
Please split this PR up into two PRs: the first with the circuit change (where as previously discussed, each commit should pass on its own including any changes to the pinned verification key), and the second adding the benchmarking.

File: src/circuit/gadget/poseidon/pow5.rs [hunk 1]
```diff
@@ -87,123 +87,96 @@ impl<F: FieldExt> Pow5T3Chip<F> {
...
-            vec![
-                (
-                    "state[0] round a",
+            let next = |idx: usize, meta: &mut VirtualCells<F>| {
+                let next_0 = meta.query_advice(state[0], Rotation::next());
+                let next_0 = next_0 * m_inv[idx][0];
+                (1..WIDTH).fold(next_0, |acc, next_idx| {
+                    let next = meta.query_advice(state[next_idx], Rotation::next());
+                    acc + next * m_inv[idx][next_idx]
+                })
```
Comment (str4d) [inline src/circuit/gadget/poseidon/pow5.rs:130]: As this covers the end of the second round in the row, we don't need to special-case the first state word:
```suggestion
                (0..WIDTH)
                    .map(|next_idx| {
                        let next = meta.query_advice(state[next_idx], Rotation::next());
                        next * m_inv[idx][next_idx]
                    })
                    .reduce(|acc, next| acc + next)
                    .expect("WIDTH > 0")
```

File: src/circuit/gadget/poseidon/pow5.rs [hunk 2]
```diff
@@ -438,83 +420,73 @@ impl<F: FieldExt> From<CellValue<F>> for StateWord<F> {
...
-                    m[2][0] * r[0] + m[2][1] * r[1] + m[2][2] * r[2],
-                ]
-            });
+            let p_mid: Option<Vec<_>> = m
+                .iter()
+                .map(|m_i| {
+                    r.as_ref().map(|r| {
+                        r.iter()
+                            .enumerate()
+                            .fold(F::zero(), |acc, (j, r_j)| acc + m_i[j] * r_j)
```
Comment (str4d) [inline src/circuit/gadget/poseidon/pow5.rs:486]:
```suggestion
                        m_i.iter()
                            .zip(r.iter())
                            .fold(F::zero(), |acc, (m_ij, r_j)| acc + *m_ij * r_j)
```

File: src/circuit/gadget/poseidon/pow5.rs [hunk 3]
```diff
@@ -529,31 +501,36 @@ impl<F: FieldExt> Pow5T3State<F> {
...
-                    r_mid.map(|r| m[2][0] * r[0] + m[2][1] * r[1] + m[2][2] * r[2]),
-                ],
-            ))
+            let state: Vec<Option<_>> = m
+                .iter()
+                .map(|m_i| {
+                    r_mid.as_ref().map(|r| {
+                        r.iter()
+                            .enumerate()
+                            .fold(F::zero(), |acc, (j, r_j)| acc + m_i[j] * r_j)
```
Comment (str4d) [inline src/circuit/gadget/poseidon/pow5.rs:519]:
```suggestion
                        m_i.iter()
                            .zip(r.iter())
                            .fold(F::zero(), |acc, (m_ij, r_j)| acc + *m_ij * r_j)
```

**Review by str4d — APPROVED** [review]
Tested ACK with suggestions. I checked locally that the commit passes tests, and that my suggestions would also do so.

**Review by str4d — APPROVED** [review]
re-ACK. PR no longer alters the circuit.

File: src/circuit/gadget/poseidon/pow5.rs [hunk 4]
```diff
@@ -48,14 +50,15 @@ impl<F: FieldExt> Pow5T3Chip<F> {
     // TODO: Does the rate need to be hard-coded here, or only the width? It probably
     // needs to be known wherever we implement the hashing gadget, but it isn't strictly
     // necessary for the permutation.
```
Comment (daira) [inline src/circuit/gadget/poseidon/pow5.rs:52]:
```suggestion
    // Currently only the case `RATE == WIDTH - 1` (i.e. with a capacity of one field
    // element) is supported.
```

File: src/circuit/gadget/poseidon/pow5.rs [hunk 5]
```diff
@@ -48,14 +47,15 @@ impl<F: FieldExt> Pow5T3Chip<F> {
...
+    pub fn configure<S: Spec<F, WIDTH, RATE>>(
         meta: &mut ConstraintSystem<F>,
         spec: S,
         state: [Column<Advice>; WIDTH],
         partial_sbox: Column<Advice>,
         rc_a: [Column<Fixed>; WIDTH],
         rc_b: [Column<Fixed>; WIDTH],
-    ) -> Pow5T3Config<F> {
+    ) -> Pow5Config<F, WIDTH, RATE> {
+        assert_eq!(RATE, WIDTH - 1);
```
[the remark this answers is not in the document]
Comment (daira) [reply src/circuit/gadget/poseidon/pow5.rs:61]:
```suggestion
        const_assert!(RATE == WIDTH - 1);
```

**Review by ebfull — APPROVED** [review]

File: src/circuit/gadget/poseidon/pow5.rs [hunk 6]
```diff
@@ -286,46 +268,52 @@ impl<F: FieldExt, S: Spec<F, WIDTH, 2>> PoseidonInstructions<F, S, WIDTH, 2> for
...
-        domain: &impl Domain<F, WIDTH, 2>,
+        domain: &impl Domain<F, WIDTH, RATE>,
     ) -> Result<State<Self::Word, WIDTH>, Error> {
         let config = self.config();
-        layouter.assign_region(
+        let state = layouter.assign_region(
             || format!("initial state for domain {:?}", domain),
             |mut region| {
-                let mut load_state_word = |i: usize, value: F| {
+                let mut state = Vec::with_capacity(WIDTH);
```
Comment (daira) [inline src/circuit/gadget/poseidon/pow5.rs:283]: The collision of the term "capacity" is unfortunate here, but it's fine.

File: src/circuit/gadget/poseidon/pow5.rs [hunk 7]
```diff
@@ -399,17 +386,19 @@ impl<F: FieldExt, S: Spec<F, WIDTH, 2>> PoseidonDuplexInstructions<F, S, WIDTH,
...
 
-    fn get_output(state: &State<Self::Word, WIDTH>) -> SpongeState<Self::Word, 2> {
-        [Some(state[0]), Some(state[1])]
+    fn get_output(state: &State<Self::Word, WIDTH>) -> SpongeState<Self::Word, RATE> {
+        state[..RATE]
+            .iter()
+            .map(|word| Some(*word))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap()
```
Comment (daira) [inline src/circuit/gadget/poseidon/pow5.rs:401]: `array::map`, which is [stable in Rust 1.55](https://github.com/rust-lang/rust/commit/7c5cabe30f4c6402e06ab0240cdfc066e7c70e8e#diff-534d866e6adad8a1f7577c2f1b3eed0b05d5f001c0d5032ac226393736940381R312), would be so much nicer!

librustzcash has a [rust-toolchain](https://github.com/zcash/librustzcash/blob/master/rust-toolchain) for Rust 1.51, but zcashd [pins Rust 1.56.1](https://github.com/zcash/zcash/blob/master/depends/packages/native_rust.mk#L2).

File: src/circuit/gadget/poseidon/pow5.rs [hunk 8]
```diff
@@ -438,83 +427,73 @@ impl<F: FieldExt> From<CellValue<F>> for StateWord<F> {
...
-            let r_0 = q_0.map(|v| v.pow(&config.alpha));
-            let r_1 = q_1.map(|v| v.pow(&config.alpha));
-            let r_2 = q_2.map(|v| v.pow(&config.alpha));
-
+            let q = self
+                .0
+                .iter()
+                .enumerate()
+                .map(|(idx, word)| word.value.map(|v| v + config.round_constants[round][idx]));
+            let r: Option<Vec<F>> = q.map(|q| q.map(|q| q.pow(&config.alpha))).collect();
```
Comment (daira) [inline src/circuit/gadget/poseidon/pow5.rs:446]:
```suggestion
            let r: Option<Vec<F>> = q.map(|qi| qi.map(|v| v.pow(&config.alpha))).collect();
```

File: src/circuit/gadget/poseidon/pow5.rs [hunk 9]
```diff
@@ -438,83 +427,73 @@ impl<F: FieldExt> From<CellValue<F>> for StateWord<F> {
...
-
-            let r = p.map(|p| {
-                [
-                    (p[0] + config.round_constants[round][0]).pow(&config.alpha),
-                    p[1] + config.round_constants[round][1],
-                    p[2] + config.round_constants[round][2],
-                ]
+            let p: Option<Vec<_>> = self.0.iter().map(|word| word.value).collect();
+
+            let r: Option<Vec<_>> = p.map(|p| {
```
Comment (daira) [inline src/circuit/gadget/poseidon/pow5.rs:471]: Again I don't particularly like the shadowing in `p.map(|p| ...)`.

**Review by daira — APPROVED** [review]
utACK with comments.

File: src/circuit/gadget/poseidon/pow5.rs (hunk 9 above)
Comment (str4d) [reply src/circuit/gadget/poseidon/pow5.rs:471]: I can try and address this in the rewrite.

File: src/circuit/gadget/poseidon/pow5.rs (hunk 7 above)
Comment (str4d) [reply src/circuit/gadget/poseidon/pow5.rs:401]: Yep, I plan to migrate everything to that at some point, but 1.51 is a reasonable schelling point for now, matching most of our dependencies. I think 1.56 will be the next MSRV bump (matching what it looks like the RustCrypto ecosystem will be moving to).

File: src/circuit/gadget/poseidon/pow5.rs (hunk 4 above)
Comment (str4d) [reply src/circuit/gadget/poseidon/pow5.rs:52]: I'll update this in the rewrite PR.

**Pull request merged**